### PR TITLE
GEP(1958): TLS ALPN Routing

### LINKED ISSUE #1958: GEP: TLS ALPN Routing
**What would you like to be added**:
A way to configure routing decisions based on the application protocols present in the TLS client hello of a TLS connection.

The `TLSRoute` is a good candidate for supporting this as it is a similar feature to SNI routing which is already handled by `TLSRoute`. Both SNI and ALPN are present in the TLS client hello and allow for TLS passthrough.

The initial discussion on this subject is [here](https://github.com/kubernetes-sigs/gateway-api/discussions/1683).

**Why this is needed**:

This would allow service owners to route to different backends behind a single domain based on the protocol the client supports.

For example if you have http/2 and mqtt clients connecting to `example.com:443` you could use the application protocols in the TLS client hello to route these clients to different backends that are able to handle the respective protocol.

### PR DESCRIPTION
**Pull request by dboslee**
**What type of PR is this?**
/kind gep

**What this PR does / why we need it**:
This PR kicks off the discussion of goals and non-goals for the GEP around TLS ALPN routing. I have left out any implementation details until there is agreement on the goals.

**Which issue(s) this PR fixes**:

Fixes #1958 
[Initial discussion](https://github.com/kubernetes-sigs/gateway-api/discussions/1683)

### CODE REVIEW

File: geps/gep-1958.md
Comment (howardjohn) [inline geps/gep-1958.md:7]: One thing we may need to take into account is not just routing based on the ALPN, but what ALPNs we should offer in the first place.

It doesn't help to have a ALPN route for `foobar` if the Gateway doesn't advertise it

Comment (dboslee) [reply geps/gep-1958.md:7]: Good point, I think that depends on whether TLS termination or TLS passthrough is being used.

I was only considering the passthrough case where the gateway is not responsible for the response. In this case the gateway is only making routing decisions, its up to the service owner to ensure the given protocol is actually selected by the upstream service.

For the TLS termination case I suppose the gateway would need to be configured to route and select ALPN based on the values configured by the service owner.

Comment (howardjohn) [reply geps/gep-1958.md:7]: Ah right, forgot that. I think TLS route is only for passthrough today (?) so that would solve that concern

File: geps/gep-1958.md
Comment (shaneutt) [inline geps/gep-1958.md:5]:
```suggestion
Add support for routing traffic based on the application protocols present in the "client hello" step of a TLS handshake.
```

File: geps/gep-1958.md
Comment (shaneutt) [inline geps/gep-1958.md:10]: :+1:

File: geps/gep-1958.md
Comment (shaneutt) [inline geps/gep-1958.md:14]: :+1:

**Review by shaneutt — APPROVED** [review]
I very much appreciate you focusing on the "what" and "why" for this initial iteration. As `Provisional` and because I see the value in the goals I approve and support moving this forward into the next steps.

Give it a bit of time to let other community members take a look:

/hold

And make sure we get at least one other maintainer to take a look:

/cc @robscott @youngnick

File: geps/gep-1958.md
Comment (costinm) [reply geps/gep-1958.md:7]: I suspect we need ALPN both for TLS passthrough and for termination, and likely the APIs will be quite different. 

For passthrough - would be good to have some example use cases, I have not seen any server doing this and I'm not sure what will the user do - for example if alpn is 'h2', would the stream be routed to a different backend ? Doesn't seem a common enough use case.

 For 'terminate' - beyond an ordered list of protocols that the server accepts, not sure what else we can do. Server doesn't advertise protocols - it needs to selects one from the list sent by the client.

Besides h2 - I have not seen a lot of broadly used ALPNs on the internet.

**Review by arkodg — APPROVED** [review]
looks good, thanks !

File: geps/gep-1958.md
Comment (howardjohn) [inline geps/gep-1958.md:23]: When we are routing it's based on the *client provided list of accepted ALPN*,right?

This may be odd for passthrough. Cliejt may present h2, http1. Route matches "h2" and routes. Backend selects "http1" .

Now our route that matched "h2" ends up serving http1/1

Comment (dboslee) [reply geps/gep-1958.md:23]: This seems fine IMO as long as long as long as its clear how routing decisions are being made.

File: geps/gep-1958.md
Comment (dboslee) [reply geps/gep-1958.md:7]: I think the two use cases and APIs are different enough that it makes sense to limit the scope this GEP to only include the tls passthrough use case.

**Comment by shaneutt** [comment]
@dboslee checking in on this one: where are we at? do you need any assistance to move this one forward?

**Comment by dboslee** [comment]
> @dboslee checking in on this one: where are we at? do you need any assistance to move this one forward?

@shaneutt The only clarification needed is on https://github.com/kubernetes-sigs/gateway-api/pull/1974#discussion_r1210574157. I would like to add the tls termination use case as a non-goal.

**Comment by costinm** [comment]
I think I'm more ok with it for tls routing - termination cases are far more complicated.

But I'm still not convinced this belong here - and not in a vendor extension. It's such an arcane and rare use case.
I assume it'll be optional - how many implementations do you expect to bother to implement it ( or have the ability 
to do so) ? Even plain SNI routing is not so common - and in some cases it gets miss-used. 

We really need a repository for extensions and uncommon features - keeping the core APIs simple would 
be better for users. 

But I'm not against it - I can see some use cases and in the case of pass-through there are very few things we can 
use in ClientHello. I can think of few ways to take advantage of this feature, after all it's a list of opaque strings...

**Comment by youngnick** [comment]
Once #2689 merges, this will need a rebase and update - the GEP files have moved.

File: geps/gep-1958.md
Comment (tokers) [inline geps/gep-1958.md:1]: That's an interesting idea, but could you share some real cases that we do need route connections according to the ALPN?

File: geps/gep-1958.md
Comment (shaneutt) [reply geps/gep-1958.md:7]: Given the lack of response, we'll consider this comment resolved for the moment. Please mark it unresolved if more conversation is needed.

File: geps/gep-1958.md
Comment (shaneutt) [reply geps/gep-1958.md:1]: @dboslee just wanted to ping on this, as I think capturing a couple "user stories" or user-story-like things in the GEP could be helpful for future readers.

File: geps/gep-1958.md
Comment (howardjohn) [inline geps/gep-1958.md:10]: This PR does not do this (or not completely, rather). Is it intended to achieve this goal in this PR or in the future?

Specifically, the handling of the fact a request has multiple ALPNs which are negotiated by the backends, and we may have listeners.

For instance, what is the behavior of:

Client ALPN: `[a, b, c]`
Listeners: `[no ALPN match, match ALPN=b, match ALPN=c]`
Server selects ALPN `c`

And many similar cases

**Pull request closed without merging**
